clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -850,54 +954,54 @@ synchronized boolean hasOngoingTransaction() {
     }
 
     synchronized boolean canRetry(ProduceResponse.PartitionResponse response, ProducerBatch batch) {
-        if (!hasProducerIdAndEpoch(batch.producerId(), batch.producerEpoch()))
-            return false;
-
         Errors error = response.error;
-        if (error == Errors.OUT_OF_ORDER_SEQUENCE_NUMBER && !hasUnresolvedSequence(batch.topicPartition) &&
-                (batch.sequenceHasBeenReset() || !isNextSequence(batch.topicPartition, batch.baseSequence())))
-            // We should retry the OutOfOrderSequenceException if the batch is _not_ the next batch, ie. its base
-            // sequence isn't the lastAckedSequence + 1. However, if the first in flight batch fails fatally, we will
-            // adjust the sequences of the other inflight batches to account for the 'loss' of the sequence range in
-            // the batch which failed. In this case, an inflight batch will have a base sequence which is
-            // the lastAckedSequence + 1 after adjustment. When this batch fails with an OutOfOrderSequence, we want to retry it.
-            // To account for the latter case, we check whether the sequence has been reset since the last drain.
-            // If it has, we will retry it anyway.
-            return true;
 
-        if (error == Errors.UNKNOWN_PRODUCER_ID) {
-            if (response.logStartOffset == -1)
-                // We don't know the log start offset with this response. We should just retry the request until we get it.
-                // The UNKNOWN_PRODUCER_ID error code was added along with the new ProduceResponse which includes the
-                // logStartOffset. So the '-1' sentinel is not for backward compatibility. Instead, it is possible for
-                // a broker to not know the logStartOffset at when it is returning the response because the partition
-                // may have moved away from the broker from the time the error was initially raised to the time the
-                // response was being constructed. In these cases, we should just retry the request: we are guaranteed
-                // to eventually get a logStartOffset once things settle down.
-                return true;
-
-            if (batch.sequenceHasBeenReset()) {
-                // When the first inflight batch fails due to the truncation case, then the sequences of all the other
-                // in flight batches would have been restarted from the beginning. However, when those responses
-                // come back from the broker, they would also come with an UNKNOWN_PRODUCER_ID error. In this case, we should not
-                // reset the sequence numbers to the beginning.
+        // For an UNKNOWN_PRODUCER_ID, the idempotent producer will locally bump the epoch and reset the sequence
```

Comment:
Hmm.. For the transactional case, the old code was able to handle UNKNOWN_PRODUCER_ID by checking the log start offset against the last acked offset. Unless I'm missing something, we seem to have lost that.